trieste/models/keras/sampler.py

```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
+        # [DAV] *** ValueError: Input 0 of layer model_2_dense_0 is incompatible with the layer: expected axis -1 of input shape to have value 1 but received input with shape (4, 2)
+        predictions = []
+        batch_index = tf.range(0, self._batch_size, 1)
+        if self._use_samples: 
+            for b, seed in zip(batch_index, tf.unstack(self._seeds)):
+                predictions.append(ensemble_distributions[self._indices[b]].sample(seed=seed))
+        else:
+            for b in batch_index:
+                predictions.append(ensemble_distributions[self._indices[b]].mean())
+        tensor_predictions = tf.squeeze(
+            tf.convert_to_tensor([unflatten(p) for p in predictions]), axis=-1
+        )  # [B, N, B]
+
+        # here we select simultaneously networks and batch dimension according to batch indices
+        # this is needed because we compute a whole batch with each network
+        indices = tf.stack([batch_index, batch_index], axis=1)
+        batch_predictions = tf.gather_nd(tf.transpose(tensor_predictions, perm=[0,2,1]), indices)  # [B,N]
+
+        return tf.transpose(batch_predictions, perm=[1, 0])  # [N, B]
+
+    def resample(self) -> None:
+        """
+        Efficiently resample network indices, and optionally quantiles, in-place without retracing.
+        """
+        self._indices.assign(self._model.sample_index(self._batch_size))  # [B]
+        if self._use_samples:
+            self._seeds.assign(tf.random.uniform(shape=(self._batch_size, 2), minval=1, maxval=999999999, dtype=tf.int32))
+
+
+class DropoutTrajectorySampler(TrajectorySampler[TrainableProbabilisticModel]):
+    """
+    This class builds functions that approximate a trajectory by using the predicted means
+    of the stochastic forward passes as a trajectory.
+    """
+
+    def __init__(self, model: TrainableProbabilisticModel):
+        """
+        :param model: The Monte Carlo Dropout model to sample from.
+        """
+        if not isinstance(model, TrainableProbabilisticModel):
+            raise NotImplementedError(
+                f"DropoutTrajectorySampler only works with Probabilistic Models; "
+                f"received {model.__repr__()}"
+            )
+
+        super().__init__(model)
+
+        self._model = model
+
+    def __repr__(self) -> str:
+        """"""
+        return f"{self.__class__.__name__}({self._model!r}"
+
+    def get_trajectory(self) -> TrajectoryFunction:
+        """
+        Generate an approximate function draw (trajectory) by using the predicted means
+    of the stochastic forward passes as a trajectory.
+
+        :return: A trajectory function representing an approximate trajectory
+            from the model, taking an input of shape `[N, 1, D]` and returning shape `[N, 1]`.
+        """
+
+        return dropout_trajectory(self._model)
+
+    def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
```

Comment:
you'll need `update_trajectory` method as well, you can copy the one from the deep ensemble sampler I think